torch/csrc/dynamo/eval_frame.c

```diff
@@ -798,19 +748,86 @@ inline static PyObject* eval_custom_code_impl(
 
   PyObject** fastlocals_old = frame->f_localsplus;
   PyObject** fastlocals_new = shadow->f_localsplus;
+  Py_ssize_t n_old = frame->f_code->co_nlocals + PyCode_GetNFreevars(frame->f_code) + PyCode_GetNCellvars(frame->f_code);
+  Py_ssize_t n_new = code->co_nlocals + PyCode_GetNFreevars(code) + PyCode_GetNCellvars(code);
+
+  #endif
 
-  for (Py_ssize_t i = 0; i < nlocals_old; i++) {
+  // ============== Initialize new frame from old frame ============
+  // Python internal for executing a function:
+  //  1. CPython interpreter first creates an empty frame according to the code object
+  //  2. CPython interpreter initializes the frame by filling arguments/free variables into frame and initializing cell variables
+  //  3. CPython interpreter executes the code object
+  // 
+  // Dynamo hooks the 3th step: before executing the code object, Dynamo transforms the code object into a new code object. Then, the old frame is not suitable for executing the new code. Therefore, Dynamo needs to manually create and initialize a new frame to execute the new code.
+  // The main task is to copy data in old frame to new frame, concerning a storage space named `localsplus`. 
+  //
+  // localsplus storage is an array with the following layout:
+  // |   args   |   new_locals    |    cell_variables |   free_variables    |
+  // | <--- from left to right, index from 0 to n - 1 ---> |
+  // code.co_varnames == args + new_locals, code.co_nlocals == len(code.co_varnames)
+  // code.co_freevars == free_variables
+  // In Python 3.10 and lower, `n == code.co_nlocals + len(code.co_cellvars) + len(code.co_freevars)` (Python expression)
+  // In Python 3.11 and higher, `n <= code.co_nlocals + len(code.co_cellvars) + len(code.co_freevars)` (Python expression). There is an extra field in Python C-API: `n == code->co_nlocalsplus` (C expression) to retrieve the length of array.
+  // The complexity happens if an argument becomes a cell variable:
+  //  In Python 3.10 and lower, `code.co_cellvars == cell_variables`, and the corresponding slot in args becomes `NULL`.
+  //  In Python 3.11 and higher, `code.co_cellvars > cell_variables`, that cell variable is still stored in args, with a flag set in corresponding item's `co_localspluskinds` .
+  //
+  // ideally, we need to look up new localsplus from old localsplus by name:
+  // for i, name, value in enumerate(localsplusnames_old):
+  //   if value != NULL: (NULL happens for new local variables and arguments that becomes cell variables)
+  //     name_to_idx[name] = i
+  // for i, name in enumerate(localsplusnames_new):
+  //  if name in name_to_idx:
+  //    fastlocals_new[i] = fastlocals_old[name_to_idx[name]]
+  // 
+  // The above process of building a `name_to_idx` mapping is expensive.
+  // Dynamo makes the following assumptions:
```

Comment:
Can we add a check for these constraints on the Python side before returning the new bytecode to the C side?

Reply:
Do you mean to check it when we have the new bytecode, rather than to check it when we run the bytecode? e.g.

https://github.com/pytorch/pytorch/blob/cc28f61fa31029466153bacc2f42a05ccdf84d45/torch/_dynamo/convert_frame.py#L567

Reply:
Yes, around there. Check when we are done creating new bytecode, not at runtime.